File: the-sweet-lab-fe/types/product.ts
export interface ProductNutrition {
    servingSize?: string;
    calories?: number;
    proteinG?: number;
    fatG?: number;
    saturatedFatG?: number;
    carbsG?: number;
    sugarG?: number;
    fiberG?: number;
    sodiumMg?: number;
    allergens?: string;
    ingredients?: string;
}

export interface Brand {
    id: string;
    name: string;
    slug: string;
    logoUrl?: string;
    description?: string;
    originCountry?: string;
    websiteUrl?: string;
}

export interface ProductVariant {
    id: string;
    name: string;
    sku: string;
    price: number;
    originalPrice?: number;
    stockQuantity: number;
    imageUrl?: string;
    displayOrder?: number;
    inStock: boolean;
}

export interface Product {
    id: string | number;
    name: string;
    slug?: string;
    sku?: string;
    price: number;
    originalPrice?: number;
    stockQuantity?: number;
    description?: string;
    thumbnailUrl?: string;
    imageUrl: string; // Tương thích hiển thị cả thumbnailUrl lẫn imageUrl
    categoryId?: string | number;
    categoryName?: string;
    categorySlug?: string;
    brand?: string;
    brandDetail?: Brand;
    origin?: string;
    cocoaPercentage?: number;
    dietaryTags?: string[];
    isFeatured?: boolean;
    isFlashSale?: boolean;
    rating: number;
    reviewCount?: number;
    soldCount?: number;
    status?: string;
    nutrition?: ProductNutrition;
    images?: string[];
    variants?: ProductVariant[];
}

export interface FlashSaleProduct {
    id: string | number;
    name: string;
    price: number;
    oldPrice: number;
    imageUrl: string;
    thumbnailUrl?: string;
    discountPercent?: number;
    rating?: number;
    reviewCount?: number;
    soldCount?: number;
    nutrition?: ProductNutrition;
    dietaryTags?: string[];
}

export interface Category {
    id: string | number;
    name: string;
    productCount: number;
    slug: string;
    backgroundColor?: string;
    imageUrl: string;
    description?: string;
    bannerUrl?: string;
    children?: Category[];
}

File: the-sweet-lab-be/src/main/java/com/thesweetlabbe/modules/product/entity/Product.java
package com.thesweetlabbe.modules.product.entity;

import com.thesweetlabbe.modules.product.enums.DietaryTag;
import com.thesweetlabbe.modules.product.enums.ProductStatus;
import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.math.BigDecimal;
import java.time.Instant;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.UUID;

@Entity
@Table(name = "products", indexes = {
        @Index(name = "idx_product_slug", columnList = "slug", unique = true),
        @Index(name = "idx_product_sku", columnList = "sku", unique = true),
        @Index(name = "idx_product_category", columnList = "category_id"),
        @Index(name = "idx_product_price", columnList = "price"),
        @Index(name = "idx_product_brand", columnList = "brand_id")
})
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Product {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @CreationTimestamp
    @Column(name = "created_at", nullable = false, updatable = false)
    private Instant createdAt;

    @UpdateTimestamp
    @Column(name = "updated_at", nullable = false)
    private Instant updatedAt;

    @Column(name = "name", nullable = false, length = 200)
    private String name;

    @Column(name = "slug", nullable = false, unique = true, length = 220)
    private String slug;

    @Column(name = "sku", nullable = false, unique = true, length = 50)
    private String sku;

    @Column(name = "price", nullable = false, precision = 15, scale = 2)
    private BigDecimal price;

    @Column(name = "original_price", precision = 15, scale = 2)
    private BigDecimal originalPrice;

    @Column(name = "stock_quantity", nullable = false)
    @Builder.Default
    private Integer stockQuantity = 0;

    @Column(name = "description", columnDefinition = "TEXT")
    private String description;

    @Column(name = "thumbnail_url", length = 500)
    private String thumbnailUrl;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "category_id", nullable = false)
    private Category category;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "brand_id")
    private Brand brand;

    @Column(name = "brand_name", length = 100)
    private String brandName;

    @Column(name = "origin", length = 100)
    private String origin;

    @Column(name = "cocoa_percentage")
    private Integer cocoaPercentage;

    @ElementCollection(targetClass = DietaryTag.class, fetch = FetchType.EAGER)
    @CollectionTable(name = "product_dietary_tags", joinColumns = @JoinColumn(name = "product_id"))
    @Enumerated(EnumType.STRING)
    @Column(name = "dietary_tag", length = 30)
    @Builder.Default
    private Set<DietaryTag> dietaryTags = new HashSet<>();

    @Column(name = "is_featured", nullable = false)
    @Builder.Default
    private boolean isFeatured = false;

    @Column(name = "is_flash_sale", nullable = false)
    @Builder.Default
    private boolean isFlashSale = false;

    @Column(name = "rating")
    @Builder.Default
    private Double rating = 5.0;

    @Enumerated(EnumType.STRING)
    @Column(name = "status", nullable = false, length = 30)
    @Builder.Default
    private ProductStatus status = ProductStatus.ACTIVE;

    @OneToOne(mappedBy = "product", cascade = CascadeType.ALL, fetch = FetchType.LAZY)
    private ProductNutrition nutrition;

    @OneToMany(mappedBy = "product", cascade = CascadeType.ALL, orphanRemoval = true)
    @Builder.Default
    private List<ProductImage> images = new ArrayList<>();

    @OneToMany(mappedBy = "product", cascade = CascadeType.ALL, orphanRemoval = true)
    @Builder.Default
    private List<ProductVariant> variants = new ArrayList<>();

    public String getDisplayBrandName() {
        if (brand != null && brand.getName() != null) {
            return brand.getName();
        }
        return brandName;
    }
}
